server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java

```diff
@@ -275,27 +295,59 @@ private int tryParse(BytesReference data, boolean lastData) throws IOException {
                     break;
                 }
                 incrementalFromOffset = nextMarker + 1;
-                line++;
-
-                if (currentRequest == null) {
-                    if (parseActionLine(data, from, nextMarker)) {
-                        if (currentRequest instanceof DeleteRequest deleteRequest) {
-                            deleteRequestConsumer.accept(deleteRequest);
-                            currentRequest = null;
-                        }
-                    }
-                } else {
-                    parseAndConsumeDocumentLine(data, from, nextMarker);
-                    currentRequest = null;
-                }
-
+                processRequest(data, from, nextMarker);
                 from = nextMarker + 1;
                 consumed = from;
             }
 
             return lastData ? from : consumed;
         }
 
+        private int tryParseWithPrefixLength(BytesReference data, boolean lastData) throws IOException {
+            int from = 0;
+            while (true) {
+                if (from == data.length()) {
+                    break;
+                }
+                if (Integer.BYTES > data.length() - from) {
+                    if (lastData) {
+                        throw new IllegalArgumentException(
+                            "Documents in the bulk request must be prefixed with the length of the document"
+                        );
+                    }
+                    break;
+                }
+                final int len = data.getInt(from);
+                if (len > data.length() - (from + Integer.BYTES)) {
+                    if (lastData) {
+                        throw new IllegalArgumentException(
+                            "Documents in the bulk request must be prefixed with the length of the document"
+                        );
+                    }
+                    break;
```

Comment:
I'm unsure of the exact control flow here so I could be wrong, but if we were to break here should `from` already be updated? since we read the bytes from `data`? Or is it deliberate to effectively leave the 4 byte length as unconsumed?